packages/headless/src/app/ssr-engine/ssr-engine.ts

```diff
@@ -0,0 +1,465 @@
+
+export type InferHydrationResult<
+  T extends EngineDefinition<
+    CoreEngine,
+    ControllerDefinitionsMap<CoreEngine, Controller>,
+    any
+  >
+> = T extends EngineDefinition<infer TEngine, infer TControllers, infer _>
+  ? EngineAndControllers<
+      TEngine,
+      InferControllersMapFromDefinition<TControllers>
+    >
+  : never;
+
+export interface ControllerSnapshot<TState> {
+  initialState: TState;
+}
+
+export interface ControllerSnapshotsMap {
+  [customName: string]: ControllerSnapshot<unknown>;
+}
+
+type AnyAction = {type: string};
+
+export interface EngineSnapshot<
+  TSearchFulfilledAction extends AnyAction,
+  TControllers extends ControllerSnapshotsMap
+> {
+  searchFulfilledAction: TSearchFulfilledAction;
+  controllers: TControllers;
+}
+
+export function mapObject<TKey extends string, TInitialValue, TNewValue>(
+  obj: Record<TKey, TInitialValue>,
+  predicate: (value: TInitialValue, key: TKey) => TNewValue
+): Record<TKey, TNewValue> {
+  return Object.fromEntries(
+    Object.entries(obj).map(([key, value]) => [
+      key,
+      predicate(value as TInitialValue, key as TKey),
+    ])
+  ) as Record<TKey, TNewValue>;
+}
+
+export function defineSearchEngine<
+  TControllerDefinitions extends ControllerDefinitionsMap<
+    SearchEngine,
+    Controller
+  >
+>({
+  controllers: controllerDefinitions,
+  ...engineOptions
+}: SearchEngineDefinitionOptions<TControllerDefinitions>): SearchEngineDefinition<TControllerDefinitions> {
+  const build: SearchEngineDefinition<TControllerDefinitions>['build'] = async (
+    ...[buildOptions]: Parameters<
+      SearchEngineDefinition<TControllerDefinitions>['build']
+    >
+  ) => {
+    const engine = buildSearchEngine(
+      buildOptions?.extend
+        ? await buildOptions.extend(engineOptions)
+        : engineOptions
+    );
+    const controllerOptions =
+      buildOptions && 'controllers' in buildOptions
+        ? buildOptions.controllers
+        : null;
+    const controllers = controllerDefinitions
+      ? mapObject(controllerDefinitions, (definition, key) =>
+          'build' in definition
+            ? definition.build(engine)
+            : definition.buildWithProps(
+                engine,
+                controllerOptions?.[key as keyof typeof controllerOptions]
+              )
+        )
+      : {};
+    return {
+      engine,
+      controllers:
+        controllers as InferControllersMapFromDefinition<TControllerDefinitions>,
+    };
+  };
+
+  const fetchInitialState: SearchEngineDefinition<TControllerDefinitions>['fetchInitialState'] =
+    (
+      ...[executeOptions]: Parameters<
+        SearchEngineDefinition<TControllerDefinitions>['fetchInitialState']
+      >
+    ) =>
+      new Promise<
+        EngineSnapshot<
+          {type: string},
+          InferControllerSnapshotsMapFromDefinitions<TControllerDefinitions>
+        >
+        // eslint-disable-next-line no-async-promise-executor
+      >(async (resolve, reject) => {
```

Comment:
I think you can replace the returned promise with an async IIFE without any params, replace the resolves by returns statement, and the reject by throw statements

Reply:
Mh true, nvm